src/device_uuid.h

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+#include <array>
+#include <stdexcept>
+#include <cstdint>
+#include <string>
+#include <vulkan/vulkan.hpp>
+
+
+template<std::size_t Size>
+constexpr std::array<char, 2 * Size> decode_UUID(const std::array<uint8_t, Size>& bytes)
+{
+    std::array<char, 2 * Size> representation{};
+    constexpr char characters[16] = 
+        { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f' };
+
+    for (std::size_t i = 0; i < bytes.size(); ++i)
+    {
+        auto& byte = bytes[i];
+
+        representation[2 * i] = characters[byte / 16];
+        representation[2 * i + 1] = characters[byte % 16];
+    }
+
+    return representation;
+}
+
+template<std::size_t Size>
+constexpr std::array<uint8_t, Size> encode_UUID(const std::array<char, 2 * Size>& representation)
+{
+    std::array<uint8_t, Size> bytes{};
+
+    auto&& decode_character = [](const char ch){
+        if (ch >= '0' && ch <= '9')
+            return ch - '0';
+        else if (ch >= 'a' && ch <= 'f')
+            return ch - 'a' + 10;
+        throw std::invalid_argument(
+            std::string{ch} + "character found while parsing hexadecimal string!");
+    };
+
+    for (std::size_t i = 0; i < bytes.size(); ++i)
+    {
+        bytes[i] = decode_character(representation[2 * i]) * 16
+            + decode_character(representation[2 * i + 1]);
+    }
+
+    return bytes;
+}
+
+struct DeviceUUID
+{
+    std::array<uint8_t, VK_UUID_SIZE> raw{};
+    
+    DeviceUUID() = default;
+    DeviceUUID(std::array<uint8_t, VK_UUID_SIZE> const& bytes)
+        : raw(bytes)
+    {}
+
+    DeviceUUID(const uint8_t bytes[])
```

Comment:
I also don't use this, but the auto compilation job on github has some old version of vulkan.hpp which returns c-like raw array (pointer) and it does not compile, and i did not want to change the call site.

![Screenshot_20210223_174719](https://user-images.githubusercontent.com/44477673/108877113-33bbfb80-75ff-11eb-8d25-8a394ee18b8c.png)

What else can i do with it?
And why is that any more unsafe than std array? Is it because the caller is now responsible to give address at exact spot where the uuid bytes begin (not a byte to the left or right)?

Reply:
It's unsafe because it doesn't enforce, at the type level, that a properly sized array is being provided (in contrast to std::array<>).

Thanks for the pointer to the CI issue, I wasn't aware of it. I checked CI and it is still using ubuntu-18.04 as the "latest" version. However, even a move to 20.04 will not fix this as the headers are still not new enough. So, I guess we can keep it for now, perhaps with a TODO comment to remove when it's not necessary any more.